### PR TITLE
NIFI-12548: Policy Management

### PR DESCRIPTION
**Pull request by mcgilman**
NIFI-12548:
- Global policy management.
- Component policy management.

### CODE REVIEW

**Comment by rfellows** [comment]
will review

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/state/access-policy/access-policy.selectors.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    let selectedResourceAction: ComponentResourceAction | null = null;
+    if (route?.params.action && route?.params.policy && route?.params.resource && route?.params.resourceIdentifier) {
+        // always select the action and resource from the route
+        selectedResourceAction = {
+            action: route.params.action,
+            policy: route.params.policy,
+            resource: route.params.resource,
+            resourceIdentifier: route.params.resourceIdentifier
+        };
+    }
```
Comment (rfellows) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/state/access-policy/access-policy.selectors.ts:64]: This is getting triggered after (or during) the component policy screen was requested to be closed. and after the state was reset. this results in the `accessPolicies/policyComponent` section of the store to get re-populated. Doesn't seem that should be happening.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/state/access-policy/access-policy.selectors.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+export const selectGlobalResourceActionFromRoute = createSelector(selectCurrentRoute, (route) => {
+    let selectedResourceAction: ResourceAction | null = null;
+    if (route?.params.action && route?.params.resource) {
+        // always select the action and resource from the route
+        selectedResourceAction = {
+            action: route.params.action,
+            resource: route.params.resource,
+            resourceIdentifier: route.params.resourceIdentifier
+        };
```
Comment (rfellows) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/state/access-policy/access-policy.selectors.ts:49]: This is getting triggered after (or during) the global policy screen was requested to be closed. and after the state was reset. this results in the `accessPolicies/accessPolicy` section of the store to get re-populated. Doesn't seem that should be happening.

**Review by rfellows — CHANGES_REQUESTED** [review]
Found a few things. I'm not quite done reviewing but I wanted to send the feedback up to this point.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/ui/component-access-policies/component-access-policies.component.ts [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+                return 'Output Ports';
+            case 'funnels':
+                return 'Funnel';
+            case 'labels':
+                return 'Label';
+            case 'remote-process-groups':
+                return 'Remote Process Group';
+        }
+
+        return 'Process Group';
```
[the remark this answers is not in the document]
Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/ui/component-access-policies/component-access-policies.component.ts:322]: Good catch!

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/feature/access-policies.component.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+})
+export class AccessPolicies implements OnInit, OnDestroy {
+    constructor(private store: Store<NiFiState>) {}
+
+    ngOnInit(): void {
+        this.store.dispatch(startCurrentUserPolling());
+    }
+
+    ngOnDestroy(): void {
+        this.store.dispatch(stopCurrentUserPolling());
```
[the remark this answers is not in the document]
Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/feature/access-policies.component.ts:36]: Good call. I'm going to add this to the global policy component as component-based policies aren't a part of the `currentUser` model.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/flow-designer/ui/canvas/header/header.component.html [hunk 5]
```diff
@@ -131,7 +131,15 @@
...
-                    <button mat-menu-item class="global-menu-item">
+                    <button
+                        mat-menu-item
+                        class="global-menu-item"
+                        [routerLink]="['/access-policies', 'global']"
+                        [disabled]="
+                            !user.tenantsPermissions.canRead ||
+                            !user.policiesPermissions.canRead ||
+                            !user.policiesPermissions.canWrite
+                        ">
```
[the remark this answers is not in the document]
Comment (mcgilman) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/flow-designer/ui/canvas/header/header.component.html:142]: Good catch. This was an oversight.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/ui/global-access-policies/global-access-policies-routing.module.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+const routes: Routes = [
+    {
+        path: ':action/:resource',
+        canMatch: [
+            authorizationGuard(
+                (user: CurrentUser) =>
+                    user.tenantsPermissions.canRead &&
+                    user.policiesPermissions.canRead &&
+                    user.policiesPermissions.canWrite
+            )
```
Comment (rfellows) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/access-policies/ui/global-access-policies/global-access-policies-routing.module.ts:33]: Do we also need to inject the FlowConfiguration here to verify `supportsManagedAuthorizer`?
If the user has the url, http://localhost:4200/#/access-policies/global/read/parameter-contexts for example, they can get to the page even though the menu options are not available.

Same is true for the users page as well.

**Review by rfellows — CHANGES_REQUESTED** [review]
Just a couple more things

**Review by rfellows — APPROVED** [review]
Looks great @mcgilman

**Pull request merged**